src/axom/sina/include/Datum.hpp

```diff
@@ -0,0 +1,198 @@
+#ifndef SINA_DATUM_HPP
+#define SINA_DATUM_HPP
+
+/// @file
+
+#include <string>
+#include <type_traits>
+
+#include "axom/sina/include/ConduitUtil.hpp"
+#include "conduit.hpp"
+
+namespace axom
+{
+namespace sina
+{
+
+/**
+ * Represents whether a Datum is a String, Scalar (double), array of Strings,
+ * or array of Scalars.
+ */
+enum class ValueType {
+    String,
+    Scalar,
+    StringArray,
+    ScalarArray
+};
+
+/**
+ * A Datum tracks the value and (optionally) tags and/or units of a
+ * value associated with a Record, e.g. a scalar, a piece of metadata,
+ * or an input parameter. In the Sina schema, a Datum always
+ * belongs to a Record or one of Record's inheriting types.
+ *
+ * Every Datum must have a value; units and tags are optional.
+ *
+ * The value of a Datum may be a string, a double, an array of strings,
+ * or an array of doubles.
+ *
+ * \code
+ * axom::sina::Datum myDatum{12.34};
+ * std::string value = "foobar";
+ * axom::sina::Datum myOtherDatum{value};
+ * std::vector<double> scalars = {1, 2, 20.0};
+ * axom::sina::Datum myArrayDatum{scalars};
+ *
+ * //prints 1, corresponding to Scalar
+ * std::cout << static_cast<std::underlying_type<sina::ValueType>::type>(myDatum.getType()) << std::endl;
+ *
+ * //prints 0, corresponding to String
+ * std::cout << static_cast<std::underlying_type<sina::ValueType>::type>(myOtherDatum.getType()) << std::endl;
+ *
+ * //prints 3, corresponding to ScalarArray
+ * std::cout << static_cast<std::underlying_type<sina::ValueType>::type>(myArrayDatum.getType()) << std::endl;
+ *
+ * myRecord->add(myDatum);
+ * myOtherDatum.setUnits("km/s");
+ * myRecord->add(myOtherDatum);
+ * std::vector<std:string> tags = {"input", "core"};
+ * myArrayDatum.setTags(tags);
+ * myRecord->add(myArrayDatum);
+ * \endcode
+ */
+class Datum {
+public:
+    /**
+     * Construct a new Datum.
+     *
+     * @param value the string value of the datum
+     */
+    Datum(std::string value);
+
+    /**
+     * Construct a new Datum.
+     *
+     * @param value the double value of the datum
+     */
+    Datum(double value);
+
+    /**
+     * Construct a new Datum.
+     *
+     * @param value the string array value of the datum
+     */
+    Datum(std::vector<std::string> value);
+
+    /**
+     * Construct a new Datum.
+     *
+     * @param value the scalar array value of the datum
+     */
+    Datum(std::vector<double> value);
+
+    /**
+     * Construct a Datum from its Node representation.
+     *
+     * @param asNode the Datum as conduit Node
+     */
+    explicit Datum(conduit::Node const &asNode);
```

Comment:
Axom tends towards "west coast" const rather than having const on the right hand side of the type. Do we need a new rule for the "make style" check? I'm assuming this has already been "make-style'd".

Reply:
Actually, the indentation would be probably be different if "make style" was run.

Reply:
@BradWhitlock we will need to pull this PR branch into the Axom repo so all the checks will run (i.e., GitLab CI). I think the make style check only runs on one of the GitLab CI test jobs.  Are you comfortable pulling in the branch and making a companion PR that is x-referenced with this one?  The instructions for that are here: https://axom.readthedocs.io/en/develop/docs/sphinx/dev_guide/dev_summary.html#contributors-and-project-access

Reply:
@rhornung67 - sure, I can do that.

Reply:
@LLNL/axom Ok I've now ran `make style` which shows:
```
Built target clangformat_style
Built target style
```
It didn't look like anything really happened here besides building executables so I then tried `make clang_tidy_style` which gave me:
```
[ 25%] Running default clang-tidy source code static analysis checks.
Enabled checks:
    modernize-use-using
    readability-braces-around-statements

clang-apply-replacements version 14.0.6
Applying fixes ...
[ 25%] Built target tools_clang_tidy_style
[ 50%] Running default clang-tidy source code static analysis checks.
Enabled checks:
    modernize-use-using
    readability-braces-around-statements

clang-apply-replacements version 14.0.6
Applying fixes ...
[ 50%] Built target thirdparty_tests_clang_tidy_style
[ 75%] Running default clang-tidy source code static analysis checks.
Enabled checks:
    modernize-use-using
    readability-braces-around-statements

clang-apply-replacements version 14.0.6
Applying fixes ...
[ 75%] Built target core_clang_tidy_style
[100%] Running default clang-tidy source code static analysis checks.
Enabled checks:
    modernize-use-using
    readability-braces-around-statements

clang-apply-replacements version 14.0.6
Applying fixes ...
[100%] Built target sina_clang_tidy_style
[100%] Built target clang_tidy_style
```
Even though it says "Applying fixes ..." no changes were made to my branch. I've also tried `make clangformat_style` and again nothing happens besides a "Built target clangformat_style" message. I feel like I'm doing something wrong here; is there a different way to run the style formatting scripts?

Reply:
Can you run the following:

```
make help | grep sina"
```

It should give you all targets related specifically to sina.

I am reworking our code checks currently to remove the component specific ones but this will currently work.

Reply:
@white238 the output of that is:
```
... sina_clang_tidy_check
... sina_clang_tidy_style
... sina
... sina_AdiakWriter_test
... sina_ConduitUtil_test
... sina_CurveSet_test
... sina_Curve_test
... sina_DataHolder_test
... sina_Datum_test
... sina_Document_test
... sina_File_test
... sina_ID_test
... sina_Record_test
... sina_Relationship_test
... sina_Run_test
... sina_basic_ex
... sina_check_datum_type_ex
... sina_create_datum_ex
... sina_curve_set_ex
... sina_document_assembly_ex
... sina_file_object_creation_ex
... sina_file_object_removal_ex
... sina_fortran_ex
... sina_local_id_relationship_ex
... sina_query_record_for_files_ex
... sina_query_records_relationships_ex
... sina_relationship_assembly_ex
... sina_set_datum_units_tags_ex
... sina_test_utils
... sina_tutorial_ex
... sina_view_datum_types_ex
... sina_view_datum_values_ex
```
I've tried running `make sina_clang_tidy_style` and again nothing is modified.

Reply:
I wonder if the current style doesn't that many nested directories because you are calling `axom_add_code_checks(PREFIX sina)`.

Gimme a sec.

Reply:
Figured out the issue.